File: src/Nav/compressed_telemetry/compressed_telemetry/costmap_compressor.py
import rclpy
from rclpy.node import Node
from rclpy.qos import QoSProfile, ReliabilityPolicy, DurabilityPolicy, HistoryPolicy
from nav_msgs.msg import OccupancyGrid
from map_msgs.msg import OccupancyGridUpdate
from std_msgs.msg import UInt8MultiArray
from rclpy.serialization import serialize_message
import zlib
import array


class CostmapCompressor(Node):
    def __init__(self):
        super().__init__("costmap_compressor")

        self.declare_parameter("log_compression_stats", True)
        self.log_compression_stats = self.get_parameter("log_compression_stats").value

        # Stats window for bandwidth/ratio measurement over 10s
        self.declare_parameter("log_mbps_stats", True)
        self.log_mbps_stats = self.get_parameter("log_mbps_stats").value

        self._raw_bytes_window = 0
        self._compressed_bytes_window = 0

        self._stats_timer = self.create_timer(10.0, self._stats_timer_callback)

        radio_qos = QoSProfile(
            reliability=ReliabilityPolicy.BEST_EFFORT,
            durability=DurabilityPolicy.VOLATILE,
            history=HistoryPolicy.KEEP_LAST,
            depth=1,
        )

        nav2_qos = QoSProfile(
            reliability=ReliabilityPolicy.RELIABLE,
            durability=DurabilityPolicy.TRANSIENT_LOCAL,
            history=HistoryPolicy.KEEP_LAST,
            depth=1,
        )

        # --- Publishers (To Radio Link) ---
        self.full_map_pub = self.create_publisher(
            UInt8MultiArray, "/telemetry/costmap_full_compressed", radio_qos
        )
        self.update_pub = self.create_publisher(
            UInt8MultiArray, "/telemetry/costmap_updates_compressed", radio_qos
        )

        # --- Subscribers (From Local Nav2) ---
        self.map_sub = self.create_subscription(
            OccupancyGrid, "/global_costmap/costmap", self.map_callback, nav2_qos
        )
        self.update_sub = self.create_subscription(
            OccupancyGridUpdate,
            "/global_costmap/costmap_updates",
            self.update_callback,
            10,
        )

        # --- State & Timers ---
        self.latest_full_map = None
        # Heartbeat timer: Send a full map refresh every 10 seconds (0.1 Hz)
        self.heartbeat_timer = self.create_timer(10.0, self.heartbeat_callback)

        self.get_logger().info("Costmap Compressor Node Started.")

    def update_callback(self, msg):
        """Instantly compress and forward high-frequency map updates."""
        try:
            raw_bytes = serialize_message(msg)
            compressed_bytes = zlib.compress(raw_bytes)

            # --- Stats Tracking ---
            raw_len = len(raw_bytes)
            compressed_len = len(compressed_bytes)
            self._raw_bytes_window += raw_len
            self._compressed_bytes_window += compressed_len
            if self.log_compression_stats:
                self._log_compression_stats("update", raw_len, compressed_len)

            out_msg = UInt8MultiArray()
            out_msg.data = array.array("B", compressed_bytes)
            self.update_pub.publish(out_msg)

        except Exception as e:
            self.get_logger().error(f"Failed to compress update: {e}")

    def map_callback(self, msg):
        """Cache the latest full map silently in the background."""
        self.latest_full_map = msg

    def heartbeat_callback(self):
        """Once every 10 seconds, send the cached full map to fix any UDP desync."""
        if self.latest_full_map is not None:
            try:
                raw_bytes = serialize_message(self.latest_full_map)
                compressed_bytes = zlib.compress(raw_bytes)

                # --- Stats Tracking ---
                raw_len = len(raw_bytes)
                compressed_len = len(compressed_bytes)
                self._raw_bytes_window += raw_len
                self._compressed_bytes_window += compressed_len

                if self.log_compression_stats:
                    self._log_compression_stats("full costmap", raw_len, compressed_len)

                out_msg = UInt8MultiArray()
                out_msg.data = array.array("B", compressed_bytes)
                self.full_map_pub.publish(out_msg)

            except Exception as e:
                self.get_logger().error(f"Failed to compress full map: {e}")

    def _log_compression_stats(self, label, raw_size, compressed_size):
        if raw_size == 0:
            self.get_logger().info(
                f"Compressed {label}: 0 bytes -> {compressed_size} bytes"
            )
            return

        ratio = compressed_size / raw_size
        reduction = (1.0 - ratio) * 100.0
        self.get_logger().info(
            f"Compressed {label}: {raw_size} bytes -> {compressed_size} bytes "
            f"({reduction:.1f}% smaller, {ratio:.2f}x of original)"
        )

    def _stats_timer_callback(self):
        if not getattr(self, "log_mbps_stats", False):
            return
        duration_s = 10.0
        compressed = float(self._compressed_bytes_window)
        raw = float(self._raw_bytes_window) if self._raw_bytes_window > 0 else 0.0
        mbps = (compressed * 8.0) / (1e6 * duration_s)

        if raw > 0.0:
            avg_ratio = compressed / raw
            avg_reduction = (1.0 - avg_ratio) * 100.0
            self.get_logger().info(
                f"Compressed send rate: {mbps:.3f} Mbps over last {int(duration_s)}s "
                f"({int(compressed)} bytes, avg reduction {avg_reduction:.1f}%, avg ratio {avg_ratio:.2f})"
            )
        else:
            self.get_logger().info(
                f"Compressed send rate: {mbps:.3f} Mbps over last {int(duration_s)}s ({int(compressed)} bytes)"
            )

        self._raw_bytes_window = 0
        self._compressed_bytes_window = 0


def main(args=None):
    rclpy.init(args=args)
    node = CostmapCompressor()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()


if __name__ == "__main__":
    main()

File: src/Nav/compressed_telemetry_cpp/src/CostmapCompressor.cpp
#include <algorithm>
#include <atomic>
#include <chrono>
#include <condition_variable>
#include <cstdint>
#include <cstring>
#include <functional>
#include <memory>
#include <mutex>
#include <string>
#include <thread>
#include <utility>
#include <vector>

#include <zlib.h>

#include "map_msgs/msg/occupancy_grid_update.hpp"
#include "nav_msgs/msg/occupancy_grid.hpp"
#include "rclcpp/rclcpp.hpp"
#include "rclcpp/serialization.hpp"
#include "rclcpp/serialized_message.hpp"
#include "rclcpp_components/register_node_macro.hpp"
#include "std_msgs/msg/u_int8_multi_array.hpp"

namespace compressed_telemetry_cpp {

namespace {
std::vector<uint8_t> CompressZlib(const uint8_t *data, size_t size) {
  uLongf compressed_size = compressBound(size);
  std::vector<uint8_t> compressed(compressed_size);
  int ret = compress2(compressed.data(), &compressed_size,
                      reinterpret_cast<const Bytef *>(data), size,
                      Z_DEFAULT_COMPRESSION);
  if (ret != Z_OK) {
    throw std::runtime_error("zlib compression failed");
  }
  compressed.resize(compressed_size);
  return compressed;
}
} // namespace

class CostmapCompressor : public rclcpp::Node {
public:
  explicit CostmapCompressor(
      const rclcpp::NodeOptions &options = rclcpp::NodeOptions())
      : rclcpp::Node("costmap_compressor", options),
        log_mbps_stats_(this->declare_parameter<bool>("log_mbps_stats", true)),
        raw_bytes_window_(0), compressed_bytes_window_(0),
        message_count_window_(0), received_count_window_(0) {
    auto radio_qos = rclcpp::QoS(1).best_effort().durability_volatile();
    auto sensor_qos = rclcpp::SensorDataQoS();

    full_map_pub_ = this->create_publisher<std_msgs::msg::UInt8MultiArray>(
        "/telemetry/costmap_full_compressed", radio_qos);
    update_pub_ = this->create_publisher<std_msgs::msg::UInt8MultiArray>(
        "/telemetry/costmap_updates_compressed", radio_qos);

    map_sub_ = this->create_subscription<nav_msgs::msg::OccupancyGrid>(
        "/global_costmap/costmap", sensor_qos,
        std::bind(&CostmapCompressor::mapCallback, this,
                  std::placeholders::_1));

    update_sub_ = this->create_subscription<map_msgs::msg::OccupancyGridUpdate>(
        "/global_costmap/costmap_updates", sensor_qos,
        std::bind(&CostmapCompressor::updateCallback, this,
                  std::placeholders::_1));

    stats_timer_ = this->create_wall_timer(
        std::chrono::seconds(20),
        std::bind(&CostmapCompressor::statsTimerCallback, this));

    map_thread_ = std::thread(&CostmapCompressor::mapProcessingLoop, this);
    update_thread_ =
        std::thread(&CostmapCompressor::updateProcessingLoop, this);

    RCLCPP_INFO(this->get_logger(), "Costmap Compressor Node Started.");
  }

  ~CostmapCompressor() override {
    {
      std::lock_guard<std::mutex> lock(map_mutex_);
      stop_processing_ = true;
    }
    map_cv_.notify_all();
    update_cv_.notify_all();
    if (map_thread_.joinable()) {
      map_thread_.join();
    }
    if (update_thread_.joinable()) {
      update_thread_.join();
    }
  }

private:
  void updateCallback(const map_msgs::msg::OccupancyGridUpdate::SharedPtr msg) {
    received_count_window_.fetch_add(1, std::memory_order_relaxed);
    {
      std::lock_guard<std::mutex> lock(update_mutex_);
      latest_update_ = msg;
    }
    update_cv_.notify_one();
  }

  void mapCallback(const nav_msgs::msg::OccupancyGrid::SharedPtr msg) {
    received_count_window_.fetch_add(1, std::memory_order_relaxed);
    {
      std::lock_guard<std::mutex> lock(map_mutex_);
      latest_full_map_ = msg;
    }
    map_cv_.notify_one();
  }

  void mapProcessingLoop() {
    while (rclcpp::ok()) {
      nav_msgs::msg::OccupancyGrid::SharedPtr current;
      {
        std::unique_lock<std::mutex> lock(map_mutex_);
        map_cv_.wait(lock,
                     [&]() { return stop_processing_ || latest_full_map_; });
        if (stop_processing_) {
          break;
        }
        current = std::move(latest_full_map_);
        latest_full_map_.reset();
      }

      if (!current) {
        continue;
      }

      try {
        rclcpp::SerializedMessage serialized_msg;
        serializer_map_.serialize_message(current.get(), &serialized_msg);
        auto &rcl_serialized = serialized_msg.get_rcl_serialized_message();
        auto compressed =
            CompressZlib(rcl_serialized.buffer, rcl_serialized.buffer_length);

        raw_bytes_window_.fetch_add(rcl_serialized.buffer_length,
                                    std::memory_order_relaxed);
        compressed_bytes_window_.fetch_add(compressed.size(),
                                           std::memory_order_relaxed);
        message_count_window_.fetch_add(1, std::memory_order_relaxed);

        std_msgs::msg::UInt8MultiArray out_msg;
        out_msg.data = std::move(compressed);
        full_map_pub_->publish(out_msg);
      } catch (const std::exception &e) {
        RCLCPP_ERROR(this->get_logger(), "Failed to compress full map: %s",
                     e.what());
      }
    }
  }

  void updateProcessingLoop() {
    while (rclcpp::ok()) {
      map_msgs::msg::OccupancyGridUpdate::SharedPtr current;
      {
        std::unique_lock<std::mutex> lock(update_mutex_);
        update_cv_.wait(lock,
                        [&]() { return stop_processing_ || latest_update_; });
        if (stop_processing_) {
          break;
        }
        current = std::move(latest_update_);
        latest_update_.reset();
      }

      if (!current) {
        continue;
      }

      try {
        rclcpp::SerializedMessage serialized_msg;
        serializer_update_.serialize_message(current.get(), &serialized_msg);
        auto &rcl_serialized = serialized_msg.get_rcl_serialized_message();
        auto compressed =
            CompressZlib(rcl_serialized.buffer, rcl_serialized.buffer_length);

        raw_bytes_window_.fetch_add(rcl_serialized.buffer_length,
                                    std::memory_order_relaxed);
        compressed_bytes_window_.fetch_add(compressed.size(),
                                           std::memory_order_relaxed);
        message_count_window_.fetch_add(1, std::memory_order_relaxed);

        std_msgs::msg::UInt8MultiArray out_msg;
        out_msg.data = std::move(compressed);
        update_pub_->publish(out_msg);
      } catch (const std::exception &e) {
        RCLCPP_ERROR(this->get_logger(), "Failed to compress update: %s",
                     e.what());
      }
    }
  }

  void statsTimerCallback() {
    if (!log_mbps_stats_) {
      return;
    }

    constexpr double duration_s = 20.0;
    const double compressed = static_cast<double>(
        compressed_bytes_window_.exchange(0, std::memory_order_relaxed));
    const double raw = static_cast<double>(
        raw_bytes_window_.exchange(0, std::memory_order_relaxed));
    const double sent_count = static_cast<double>(
        message_count_window_.exchange(0, std::memory_order_relaxed));
    const double received_count = static_cast<double>(
        received_count_window_.exchange(0, std::memory_order_relaxed));
    const double mbps = (compressed * 8.0) / (1e6 * duration_s);
    const double recv_hz = received_count / duration_s;
    const double sent_hz = sent_count / duration_s;

    if (sent_count > 0.0 && raw > 0.0) {
      const double avg_raw = raw / sent_count;
      const double avg_compressed = compressed / sent_count;
      const double avg_ratio = compressed / raw;
      const double avg_reduction = (1.0 - avg_ratio) * 100.0;
      RCLCPP_INFO(this->get_logger(),
                  "Avg compression over %d s: %.1f%% (avg %.0f bytes -> %.0f "
                  "bytes), %.3f Mbps, incoming %.2f Hz, compressed %.2f Hz",
                  static_cast<int>(duration_s), avg_reduction, avg_raw,
                  avg_compressed, mbps, recv_hz, sent_hz);
    } else {
      RCLCPP_INFO(this->get_logger(),
                  "Avg compression over %d s: no data (%.3f Mbps, incoming "
                  "%.2f Hz, compressed %.2f Hz)",
                  static_cast<int>(duration_s), mbps, recv_hz, sent_hz);
    }

    (void)sent_count;
  }

  rclcpp::Serialization<nav_msgs::msg::OccupancyGrid> serializer_map_;
  rclcpp::Serialization<map_msgs::msg::OccupancyGridUpdate> serializer_update_;

  bool log_mbps_stats_;
  std::atomic<uint64_t> raw_bytes_window_;
  std::atomic<uint64_t> compressed_bytes_window_;
  std::atomic<uint64_t> message_count_window_;
  std::atomic<uint64_t> received_count_window_;

  rclcpp::TimerBase::SharedPtr stats_timer_;
  std::mutex map_mutex_;
  std::mutex update_mutex_;
  std::condition_variable map_cv_;
  std::condition_variable update_cv_;
  nav_msgs::msg::OccupancyGrid::SharedPtr latest_full_map_;
  map_msgs::msg::OccupancyGridUpdate::SharedPtr latest_update_;
  std::thread map_thread_;
  std::thread update_thread_;
  std::atomic<bool> stop_processing_{false};

  rclcpp::Publisher<std_msgs::msg::UInt8MultiArray>::SharedPtr full_map_pub_;
  rclcpp::Publisher<std_msgs::msg::UInt8MultiArray>::SharedPtr update_pub_;
  rclcpp::Subscription<nav_msgs::msg::OccupancyGrid>::SharedPtr map_sub_;
  rclcpp::Subscription<map_msgs::msg::OccupancyGridUpdate>::SharedPtr
      update_sub_;
};

} // namespace compressed_telemetry_cpp

RCLCPP_COMPONENTS_REGISTER_NODE(compressed_telemetry_cpp::CostmapCompressor)
